src/lib/IFrameCommunicator.ts

```diff
@@ -62,26 +48,63 @@ export class IFrameCommunicator {
 
   public destroy(): void {
     this.events.removeAllListeners();
-    window.removeEventListener('message', this.onMessage, true);
+    window.removeEventListener('message', this.onMessageWithResponse, true);
   }
 
   public sendMessage = <M extends keyof IFrameMessagesMap>(
     type: M,
     data: Parameters<IFrameMessagesMap[M]>[0],
-    metadata?: Parameters<IFrameMessagesMap[M]>[1]
+    metadata?: Parameters<IFrameMessagesMap[M]>[1],
+    operationId?: string
   ): void => {
     if (!isInsideIFrame()) {
       console.error("Warning: can't send an iframe message, because the application is not hosted in an iframe");
       return;
     }
 
-    const message: IFrameMessage<M> = { type, data, metadata };
+    const message: IFrameMessage<M> = { type, data, metadata, operationId };
     window.parent.postMessage(message, '*');
   };
 
-  private onMessage = (event: MessageEvent): void => {
+  public sendMessageWithResponse = <M extends keyof IFrameMessagesMap>(
+    type: M,
+    data: Parameters<IFrameMessagesMap[M]>[0],
+    callback: () => void,
+    metadata?: Parameters<IFrameMessagesMap[M]>[1]
+  ): void => {
+    const operationId = this.createUuid();
+    this.registerCallback(callback, operationId);
+    this.sendMessage(type, data, metadata, operationId);
+  };
+
+  private onMessageWithResponse = (event: MessageEvent): void => {
     if (!event.data) return;
     const message = event.data as IFrameMessage<any>;
     this.events.emit(message.type, message.data);
+
+    if (message.operationId) {
+      this.executeCallbacks(message.operationId, {});
```

Comment:
if I got it right, `{}` should be replaced by `message.data` ?

Reply:
Also, there should be single callback per operationId, right? And the method `executeCallbacks` only executes single callback. Therefore I would call it singular, `executeCallback`. What do you think?

Reply:
This remains from the concept. I'm going to adjust this as soon as the Safelife page will run locally (need to debug)